### PR TITLE
go_1_21: install from distpack archive

### PR DESCRIPTION
**Pull request by tie**
## Description of changes

Changes the build and install phases to use archives produced by [distpack]. This allows us to drop the code that cleans GOROOT and moves cross-compiled binaries out of bin/GOOS_GOARCH, and potentially avoid issues like [missing GOROOT/go.env] in the future.

[distpack]: https://pkg.go.dev/cmd/distpack
[missing GOROOT/go.env]: https://github.com/NixOS/nixpkgs/commit/cbc976a97c3372e1eec5db021db994b85e098d12

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
Please move the finalattrs changes to a separate PR. The rebuild count for 1.21 should be low so we can send both of these changes to master.

**Comment by zowoq** [comment]
I've opened https://github.com/NixOS/nixpkgs/pull/254502 for finalAttrs. It should be zero rebuilds so I've applied it to all go versions, I've added you as co-author there.

File: pkgs/development/compilers/go/1.21.nix [hunk 1]
```diff
@@ -132,34 +131,24 @@ stdenv.mkDerivation rec {
...
-    mkdir -p $GOROOT_FINAL
-    cp -a bin pkg src lib misc api doc go.env $GOROOT_FINAL
-    mkdir -p $out/bin
-    ln -s $GOROOT_FINAL/bin/* $out/bin
+
+    mkdir -p $out/{share,bin}
+    tar -C $out/share -x -z -f "pkg/distpack/$(sed 1q VERSION).$GOOS-$GOARCH.tar.gz"
+    ln -s $out/share/go/bin/* $out/bin
+
+    cp -a pkg/distpack $dist
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:150]: Why is `pkg/distpack` being packaged as well?

File: pkgs/development/compilers/go/1.21.nix [hunk 2]
```diff
@@ -91,9 +89,8 @@ stdenv.mkDerivation rec {
 
   GOOS = stdenv.targetPlatform.parsed.kernel.name;
   GOARCH = goarch stdenv.targetPlatform;
+
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:92]: No unnecessary whitespace changes please.
```suggestion
```

File: pkgs/development/compilers/go/1.21.nix [hunk 3]
```diff
@@ -132,34 +131,24 @@ stdenv.mkDerivation rec {
     ulimit -a
 
     pushd src
-    ./make.bash
+    bash make.bash -no-banner -distpack
     popd
+
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:136]:
```suggestion

```

File: pkgs/development/compilers/go/1.21.nix [hunk 4]
```diff
@@ -132,34 +131,24 @@ stdenv.mkDerivation rec {
...
+  # NB the first line of the VERSION file is the Go version, additional lines
+  # are 'key value' pairs setting other data.
+  # See also https://pkg.go.dev/cmd/distpack
   installPhase = ''
     runHook preInstall
-    mkdir -p $GOROOT_FINAL
-    cp -a bin pkg src lib misc api doc go.env $GOROOT_FINAL
-    mkdir -p $out/bin
-    ln -s $GOROOT_FINAL/bin/* $out/bin
+
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:145]:
```suggestion
```

File: pkgs/development/compilers/go/1.21.nix [hunk 5]
```diff
@@ -132,34 +131,24 @@ stdenv.mkDerivation rec {
...
+    mkdir -p $out/{share,bin}
+    tar -C $out/share -x -z -f "pkg/distpack/$(sed 1q VERSION).$GOOS-$GOARCH.tar.gz"
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:147]: Is there a reason to read the version from the file when we already know what the version is?

```suggestion
    tar -C $out/share -x -z -f "pkg/distpack/go${version}.$GOOS-$GOARCH.tar.gz"
```

File: pkgs/development/compilers/go/1.21.nix (hunk 1 above)
Comment (tie) [reply pkgs/development/compilers/go/1.21.nix:150]: Mostly with the intent to add passthru.tests to verify that (at least) source tarball builds are identical to official releases. That is, compare `$dist/go1.21.1.src.tar.gz` to `$src`. See also https://github.com/golang/go/issues/62053

>One of the goals of Go 1.21's reproducible builds effort was to have the binaries we post on go.dev/dl match the binaries that other distributions post after building from source. Of course, to make that the case, the other distributions need to not patch the source.

Although that is not possible right now because `patches` 😅

Now that I think about it, comparing source tarball does not really make sense, especially if we eventually remove patches (e.g. by implementing required features upstream). I’ll remove `dist` output, also because files are compressed and that doesn’t really play well with Nix dependency tracking.

Comment (tie) [reply pkgs/development/compilers/go/1.21.nix]: Removed `dist` output.

File: pkgs/development/compilers/go/1.21.nix (hunk 5 above)
Comment (tie) [reply pkgs/development/compilers/go/1.21.nix:145]: Agreed, this is much better that reading `VERSION`.

Comment (tie) [reply pkgs/development/compilers/go/1.21.nix:145]: I think I copied this `sed` snippet from `cmd/distpack` documentation and forgot that we can just do `go${version}`.

File: pkgs/development/compilers/go/1.21.nix [hunk 6]
```diff
@@ -132,34 +128,19 @@ stdenv.mkDerivation rec {
...
-  '' else lib.optionalString (stdenv.hostPlatform.system != stdenv.targetPlatform.system) ''
-    rm -rf bin/*_*
-    ${lib.optionalString (!(GOHOSTARCH == GOARCH && GOOS == GOHOSTOS)) ''
-      rm -rf pkg/${GOOS}_${GOARCH} pkg/tool/${GOOS}_${GOARCH}
-    ''}
-  '');
-
+  # NB the first line of the VERSION file is the Go version, additional lines
+  # are 'key value' pairs setting other data.
+  # See also https://pkg.go.dev/cmd/distpack
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:138]: This comment isn't needed now.
```suggestion
```

File: pkgs/development/compilers/go/1.21.nix (hunk 5 above)
Comment (tie) [reply pkgs/development/compilers/go/1.21.nix]: Fixed.

File: pkgs/development/compilers/go/1.21.nix [hunk 7]
```diff
@@ -116,14 +110,16 @@ stdenv.mkDerivation rec {
 
   GOROOT_BOOTSTRAP = if useGccGoBootstrap then goBootstrap else "${goBootstrap}/share/go";
 
+  # Note that we use distpack to avoid moving around cross-compiled binaries.
+  # The paths are slightly different when buildPlatform != hostPlatform and
+  # distpack handles assembling outputs in the right place, same as the official
+  # Go binary releases.
```
Comment (zowoq) [inline pkgs/development/compilers/go/1.21.nix:116]:
```suggestion
  # Go binary releases. https://pkg.go.dev/cmd/distpack
```

File: pkgs/development/compilers/go/1.21.nix (hunk 6 above)
Comment (tie) [reply pkgs/development/compilers/go/1.21.nix]: Oops, fixed. Also moved `See also …` to the comment above `buildPhase` since this is a new tool/flag and a link may be useful.

File: pkgs/development/compilers/go/1.21.nix [hunk 8]
```diff
@@ -116,14 +110,16 @@ stdenv.mkDerivation rec {
...
+  # The paths are slightly different when buildPlatform != hostPlatform and
+  # distpack handles assembling outputs in the right place, same as the official
+  # Go binary releases. See also https://pkg.go.dev/cmd/distpack
   buildPhase = ''
     runHook preBuild
     export GOCACHE=$TMPDIR/go-cache
     # this is compiled into the binary
     export GOROOT_FINAL=$out/share/go
 
-    export PATH=$(pwd)/bin:$PATH
```
Comment (tie) [inline pkgs/development/compilers/go/1.21.nix:125]: FWIW I think this was added to silence the part of the `cmd/dist` banner that asks to add Go to the `PATH`.
https://github.com/golang/go/blob/5eb382fc08fb32592e9585f9cb99005696a38b49/src/cmd/dist/build.go#L1877-L1879

This is not needed with `-no-banner`.

File: pkgs/development/compilers/go/1.21.nix (hunk 7 above)
Comment (tie) [reply pkgs/development/compilers/go/1.21.nix]: Done.

**Comment by tie** [comment]
I’ll mark this as ready once `nixpkgs-review pr 254497` is done running on my `aarch64-linux` and `x86_64-linux` systems.

**Comment by tie** [comment]
Result of `nixpkgs-review pr 254497` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>carapace</li>
    <li>clash</li>
    <li>diffoci</li>
    <li>go_1_21</li>
    <li>juicity</li>
    <li>kubectl-klock</li>
    <li>oneshot</li>
    <li>reviewdog</li>
    <li>sq</li>
    <li>tailscale</li>
    <li>terraform-providers.tailscale</li>
    <li>trayscale</li>
    <li>trufflehog</li>
    <li>usql</li>
  </ul>
</details>

**Comment by tie** [comment]
Result of `nixpkgs-review pr --skip-package trayscale 254497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>carapace</li>
    <li>clash</li>
    <li>clash-verge</li>
    <li>diffoci</li>
    <li>go_1_21</li>
    <li>juicity</li>
    <li>kubectl-klock</li>
    <li>oneshot</li>
    <li>reviewdog</li>
    <li>sq</li>
    <li>tailscale</li>
    <li>terraform-providers.tailscale</li>
    <li>trufflehog</li>
    <li>usql</li>
  </ul>
</details>

---

[Trayscale](https://github.com/DeedleFake/trayscale) takes almost an hour to build, yikes.

**Comment by zowoq** [comment]
I've checked a couple of cross builds and it seems fine.

@nixos/golang I was going to start work on bumping the default go to 1.21 soon but I think I'll leave it another week or two so we can deal with any potential issues from this on master instead of staging.

**Comment by zowoq** [comment]
Thanks!

**Comment by saschagrunert** [comment]
Hm, this results in using the native (like arm64) golang binary instead of the system local one when cross-compiling in https://github.com/cri-o/cri-o/pull/7319 resulting in an exec binary error. Is there anything we can do in using the x86_64 golang when using

```nix
  crossSystem = {
    config = "aarch64-unknown-linux-gnu";
  };
```

**Comment by tie** [comment]
Seems to be because we set `GOOS`/`GOARCH` to targetPlatform, and before this change there was [`preInstall` phase](https://github.com/NixOS/nixpkgs/blob/e67b7b6ab5e1c8055597d5c02ad6d3c489f0f134/pkgs/development/compilers/go/1.21.nix#L140-L155) that instead kept the binaries for hostPlatform when cross-compiling.

That worked because Go builds binaries for both local and cross system and we could choose which to keep. However, Go is a cross-compiler so `targetPlatform` only makes sense for setting `CXX`/`CC_FOR_${GOOS}_${GOARCH}` where GOOS/GOARCH are inferred from `targetPlatform` (see also https://golang.org/cl/76018).

I’m working on a fix for this, I’ll make a PR soon.

**Comment by saschagrunert** [comment]
Thank you @tie!

**Comment by zowoq** [comment]
I'll revert this in https://github.com/NixOS/nixpkgs/pull/254497 for now.
